File: yandex_backend/48. Оценка разнообразия/solution_.py
import sys
from typing import Dict, List

def main() -> None:
    n: int = int(input())
    product_category: Dict[int, int] = {}
    for _ in range(n):
        product, category = map(int, input().split())
        product_category[product] = category
    
    product_order: List[int] = list(map(int, input().split()))
    
    category_positions: Dict[int, List[int]] = {}
    for i, product in enumerate(product_order):
      category: int = product_category[product]
      if category not in category_positions:
        category_positions[category] = []
      category_positions[category].append(i)
    
    if len(category_positions) == n:
       print(n)
       return

    min_diff: float = float('inf')
    for category, positions in category_positions.items():
      if len(positions) > 1:
        for i in range(len(positions) - 1):
          diff: int = positions[i+1] - positions[i]
          min_diff = min(min_diff, diff)
    
    if min_diff == float('inf'):
        print(n)
    else:
        print(min_diff)

if __name__ == '__main__':
    main()

File: yandex_backend/48. Оценка разнообразия/solution.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class DiversityCalculator
{
    public static void Main(string[] args)
    {
        int n = int.Parse(Console.ReadLine());

        Dictionary<int, int> productCategory = new Dictionary<int, int>();
        for (int i = 0; i < n; i++)
        {
            int[] input = Array.ConvertAll(Console.ReadLine().Split(' '), int.Parse);
            productCategory[input[0]] = input[1];
        }

        int[] productOrder = Array.ConvertAll(Console.ReadLine().Split(' '), int.Parse);

        Dictionary<int, List<int>> categoryPositions = new Dictionary<int, List<int>>();
        for (int i = 0; i < n; i++)
        {
            int category = productCategory[productOrder[i]];
            if (!categoryPositions.ContainsKey(category))
            {
                categoryPositions[category] = new List<int>();
            }
            categoryPositions[category].Add(i);
        }

        if (categoryPositions.Count == n)
        {
            Console.WriteLine(n);
            return;
        }

        int minDiff = int.MaxValue;
        foreach (var kvp in categoryPositions)
        {
            List<int> positions = kvp.Value;
            for (int i = 0; i < positions.Count - 1; i++)
            {
                minDiff = Math.Min(minDiff, positions[i + 1] - positions[i]);
            }
        }

        Console.WriteLine(minDiff == int.MaxValue ? n : minDiff);
    }
}